#**Tugas Kelompok 7 Machine Learning - Ensemble Learning**

In [ ]:
print("Kelompok 7")
print("Hanyel Daryus Bancin - 20/455445/PA/19660 ")
print("M. Daffa Gashandy - 20/455449/PA/19664")
print("M. Farhan Dwi Rizqi - 20/459271/PA/19932")

Kelompok 7
Hanyel Daryus Bancin - 20/455445/PA/19660 
M. Daffa Gashandy - 20/455449/PA/19664
M. Farhan Dwi Rizqi - 20/459271/PA/19932


In [ ]:
print("https://colab.research.google.com/drive/1xoCoun0m-LVqivUjST1jYwny2UB1JZoQ?usp=sharing")

https://colab.research.google.com/drive/1xoCoun0m-LVqivUjST1jYwny2UB1JZoQ?usp=sharing


#Import Library

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import sklearn
import os
import warnings
import xgboost as xgb
import lightgbm as lgb
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression, Lasso, ElasticNet
from sklearn.kernel_ridge import KernelRidge
from sklearn.ensemble import BaggingRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.ensemble import StackingRegressor
from sklearn.ensemble import AdaBoostRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.ensemble import VotingRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import mean_squared_error 
n_jobs = -1
warnings.filterwarnings("ignore") 
random_state = 69 

#Load Dataset Train

In [ ]:
# Mengupload menggunakan metode import file dari directory PC
from google.colab import files
load_datatrain = files.upload()

Saving house prices_train.csv to house prices_train.csv


In [ ]:
import io
df_train = pd.read_csv(io.BytesIO(load_datatrain['house prices_train.csv']))
df_train.head(10) #Menampilkan 10 data pertama


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
2         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
3         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
5         Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
7         Lvl    AllPub  ...        0    NaN    NaN        Shed     350   
8         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
9         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

#Pandas

b) Mendapatkan Info dataset

In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

##Deskriptif

c) Analisis Statistik Deskriptif

In [ ]:
df_train.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

# Load Dataset Test

In [ ]:
# Mengupload menggunakan metode import file dari directory PC
from google.colab import files
load_dataset = files.upload()

Saving house prices_test.csv to house prices_test.csv


In [ ]:
# Menampilkan 10 data pertama
import io
df_test = pd.read_csv(io.BytesIO(load_dataset['house prices_test.csv']))
df_test.head(10) 


Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   
5  1466          60       RL         75.0    10000   Pave   NaN      IR1   
6  1467          20       RL          NaN     7980   Pave   NaN      IR1   
7  1468          60       RL         63.0     8402   Pave   NaN      IR1   
8  1469          20       RL         85.0    10176   Pave   NaN      Reg   
9  1470          20       RL         70.0     8400   Pave   NaN      Reg   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   
5         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
6         Lvl    AllPub  ...           0        0    NaN  GdPrv        Shed   
7         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
8         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
9         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  
5       0      4    2010        WD         Normal  
6     500      3    2010        WD         Normal  
7       0      5    2010        WD         Normal  
8       0      2    2010        WD         Normal  
9       0      4    2010        WD         Normal  

[10 rows x 80 columns]

##Pandas

Mendapatkan Info Dataset Test

In [ ]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

##Deskripsi

In [ ]:
df_test.describe()

Id   MSSubClass  LotFrontage       LotArea  OverallQual  \
count  1459.000000  1459.000000  1232.000000   1459.000000  1459.000000   
mean   2190.000000    57.378341    68.580357   9819.161069     6.078821   
std     421.321334    42.746880    22.376841   4955.517327     1.436812   
min    1461.000000    20.000000    21.000000   1470.000000     1.000000   
25%    1825.500000    20.000000    58.000000   7391.000000     5.000000   
50%    2190.000000    50.000000    67.000000   9399.000000     6.000000   
75%    2554.500000    70.000000    80.000000  11517.500000     7.000000   
max    2919.000000   190.000000   200.000000  56600.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1459.000000  1459.000000   1459.000000  1444.000000  1458.000000  ...   
mean      5.553804  1971.357779   1983.662783   100.709141   439.203704  ...   
std       1.113740    30.390071     21.130467   177.625900   455.268042  ...   
min       1.000000  1879.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1963.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1992.000000     0.000000   350.500000  ...   
75%       6.000000  2001.000000   2004.000000   164.000000   753.500000  ...   
max       9.000000  2010.000000   2010.000000  1290.000000  4010.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  1458.000000  1459.000000  1459.000000    1459.000000  1459.000000   
mean    472.768861    93.174777    48.313914      24.243317     1.794380   
std     217.048611   127.744882    68.883364      67.227765    20.207842   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     318.000000     0.000000     0.000000       0.000000     0.000000   
50%     480.000000     0.000000    28.000000       0.000000     0.000000   
75%     576.000000   168.000000    72.000000       0.000000     0.000000   
max    1488.000000  1424.000000   742.000000    1012.000000   360.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  1459.000000  1459.000000   1459.000000  1459.000000  1459.000000  
mean     17.064428     1.744345     58.167923     6.104181  2007.769705  
std      56.609763    30.491646    630.806978     2.722432     1.301740  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     4.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     576.000000   800.000000  17000.000000    12.000000  2010.000000  

[8 rows x 37 columns]

#DATA PREPROCESSING

##Data Cleaning

In [ ]:
def clean_dataset(df):
    assert isinstance(df, pd.DataFrame), "df needs to be a pd.DataFrame"
    df.dropna(inplace=True)
    indices_to_keep = ~df.isin([np.nan, np.inf, -np.inf]).any(1)
    return df[indices_to_keep].astype(np.float64)

clean_dataset(scaled_dt)
clean_dataset(scaled_ds)
clean_dataset(dt)
clean_dataset(ds)

Id  MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  \
0     1461.0        20.0       2.0         80.0  11622.0     1.0    2.0   
1     1462.0        20.0       3.0         81.0  14267.0     1.0    2.0   
2     1463.0        60.0       3.0         74.0  13830.0     1.0    2.0   
3     1464.0        60.0       3.0         78.0   9978.0     1.0    2.0   
4     1465.0       120.0       3.0         43.0   5005.0     1.0    2.0   
...      ...         ...       ...          ...      ...     ...    ...   
1451  2912.0        20.0       3.0         80.0  13384.0     1.0    2.0   
1452  2913.0       160.0       4.0         21.0   1533.0     1.0    2.0   
1455  2916.0       160.0       4.0         21.0   1894.0     1.0    2.0   
1456  2917.0        20.0       3.0        160.0  20000.0     1.0    2.0   
1458  2919.0        60.0       3.0         74.0   9627.0     1.0    2.0   

      LotShape  LandContour  Utilities  ...  ScreenPorch  PoolArea  PoolQC  \
0          3.0          3.0        0.0  ...        120.0       0.0     2.0   
1          0.0          3.0        0.0  ...          0.0       0.0     2.0   
2          0.0          3.0        0.0  ...          0.0       0.0     2.0   
3          0.0          3.0        0.0  ...          0.0       0.0     2.0   
4          0.0          1.0        0.0  ...        144.0       0.0     2.0   
...        ...          ...        ...  ...          ...       ...     ...   
1451       3.0          3.0        0.0  ...          0.0       0.0     2.0   
1452       3.0          3.0        0.0  ...          0.0       0.0     2.0   
1455       3.0          3.0        0.0  ...          0.0       0.0     2.0   
1456       3.0          3.0        0.0  ...          0.0       0.0     2.0   
1458       3.0          3.0        0.0  ...          0.0       0.0     2.0   

      Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0       2.0          3.0      0.0     6.0  2010.0       8.0            4.0  
1       4.0          0.0  12500.0     6.0  2010.0       8.0            4.0  
2       2.0          3.0      0.0     3.0  2010.0       8.0            4.0  
3       4.0          3.0      0.0     6.0  2010.0       8.0            4.0  
4       4.0          3.0      0.0     1.0  2010.0       8.0            4.0  
...     ...          ...      ...     ...     ...       ...            ...  
1451    4.0          3.0      0.0     5.0  2006.0       8.0            4.0  
1452    4.0          3.0      0.0    12.0  2006.0       8.0            0.0  
1455    4.0          3.0      0.0     4.0  2006.0       8.0            0.0  
1456    4.0          3.0      0.0     9.0  2006.0       8.0            0.0  
1458    4.0          3.0      0.0    11.0  2006.0       8.0            4.0  

[1146 rows x 80 columns]

In [ ]:
x = (df_train.dtypes == 'object')
object_cols = list(x[x].index)
y = (df_test.dtypes == 'object')
object_cols1 = list(y[y].index)

print("Categorical variables in the datatrain:", object_cols)
print("Categorical variables in the datatest:", object_cols1)

Categorical variables in the datatrain: ['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']
Categorical variables in the datatest: ['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFin

##Label Encoder Data

Karena dataset berisi variabel karakter, maka harus menyandikannya menggunakan LabelEncoder

In [ ]:
#Label Encoding the object dtypes.
LE=LabelEncoder()
for i in object_cols:
    df_train[i]=df_train[[i]].apply(LE.fit_transform)

for j in object_cols1:
    df_test[j]=df_test[[j]].apply(LE.fit_transform)
    
print("All features are now numerical")

All features are now numerical


##Normalisasi(Feature scaling)

In [ ]:
#Creating a copy of data
dt = df_train.copy()
ds = df_test.copy()
#Scaling
scaler = StandardScaler()
scaler.fit(dt)
scaled_dt = pd.DataFrame(scaler.transform(dt),columns= dt.columns )
scaler.fit(ds)
scaled_ds = pd.DataFrame(scaler.transform(ds),columns= ds.columns )
print("All features are now scaled")

All features are now scaled


In [ ]:
print(f'Shape of datatrain = {scaled_dt.shape}')
scaled_dt.head(10)

Shape of datatrain = (1121, 81)


Id  MSSubClass  MSZoning  LotFrontage   LotArea    Street     Alley  \
0  -1.730865    0.073375 -0.045532    -0.208034 -0.207142  0.064238  0.244717   
1  -1.728492   -0.872563 -0.045532     0.409895 -0.091886  0.064238  0.244717   
2  -1.726120    0.073375 -0.045532    -0.084449  0.073480  0.064238  0.244717   
3  -1.723747    0.309859 -0.045532    -0.414011 -0.096897  0.064238  0.244717   
4  -1.721374    0.073375 -0.045532     0.574676  0.375148  0.064238  0.244717   
5  -1.719002   -0.163109 -0.045532     0.615871  0.360616  0.064238  0.244717   
6  -1.716629   -0.872563 -0.045532     0.203918 -0.043379  0.064238  0.244717   
8  -1.711883   -0.163109  1.537245    -0.784768 -0.440659  0.064238  0.244717   
9  -1.709511    3.147673 -0.045532    -0.825963 -0.310370  0.064238  0.244717   
10 -1.707138   -0.872563 -0.045532    -0.002058  0.068469  0.064238  0.244717   

    LotShape  LandContour  Utilities  ...  PoolArea    PoolQC     Fence  \
0   0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
1   0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
2  -1.378933     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
3  -1.378933     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
4  -1.378933     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
5  -1.378933     0.314667   -0.02618  ... -0.068692  0.063305 -1.390058   
6   0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
8   0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
9   0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   
10  0.750731     0.314667   -0.02618  ... -0.068692  0.063305  0.457447   

    MiscFeature   MiscVal    MoSold    YrSold  SaleType  SaleCondition  \
0      0.191594 -0.087688 -1.599111  0.138777  0.313867       0.208502   
1      0.191594 -0.087688 -0.489110 -0.614439  0.313867       0.208502   
2      0.191594 -0.087688  0.990891  0.138777  0.313867       0.208502   
3      0.191594 -0.087688 -1.599111 -1.367655  0.313867      -3.426284   
4      0.191594 -0.087688  2.100892  0.138777  0.313867       0.208502   
5     -4.759330  1.323736  1.360892  0.891994  0.313867       0.208502   
6      0.191594 -0.087688  0.620891 -0.614439  0.313867       0.208502   
8      0.191594 -0.087688 -0.859110  0.138777  0.313867      -3.426284   
9      0.191594 -0.087688 -1.969111  0.138777  0.313867       0.208502   
10     0.191594 -0.087688 -1.599111  0.138777  0.313867       0.208502   

    SalePrice  
0    0.347273  
1    0.007288  
2    0.536154  
3   -0.515281  
4    0.869843  
5   -0.477505  
6    1.587588  
8   -0.642461  
9   -0.792306  
10  -0.647497  

[10 rows x 81 columns]

In [ ]:
scaled_ds = scaled_ds.iloc[25:]
print(f'Shape of datatest = {scaled_ds.shape}')
scaled_ds.head(10)

Shape of datatest = (1121, 80)


Id  MSSubClass  MSZoning  LotFrontage   LotArea   Street     Alley  \
26 -1.669132    0.061351 -0.046591    -0.383604  1.536964  0.06426  0.269555   
27 -1.666758   -0.874711 -0.046591     0.197590 -0.167984  0.06426  0.269555   
28 -1.664384   -0.874711 -2.939201     1.047027 -0.121959  0.06426  0.269555   
29 -1.662009   -0.874711 -2.939201     0.689369  0.125325  0.06426  0.269555   
30 -1.659635    0.061351 -0.046591     0.063468  0.424085  0.06426  0.269555   
31 -1.657261   -0.640695 -1.492896     0.063468 -0.003868  0.06426  0.269555   
32 -1.654886   -0.874711 -0.046591    -1.322455  1.128592  0.06426  0.269555   
33 -1.652512    0.061351 -0.046591     0.734076  0.670965  0.06426  0.269555   
34 -1.650138    0.061351 -0.046591     0.868198  0.265419  0.06426  0.269555   
35 -1.647764    1.465443 -2.939201    -1.948356 -1.006125  0.06426  0.269555   

    LotShape  LandContour  Utilities  ...  ScreenPorch  PoolArea    PoolQC  \
26 -0.675551       0.3187   -0.03705  ...     4.222196 -0.057227  0.043675   
27  0.742620       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
28  0.742620       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
29  0.742620       0.3187   -0.03705  ...     3.515362 -0.057227  0.043675   
30  0.742620       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
31  0.742620       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
32 -1.384637       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
33 -1.384637       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
34  0.742620       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   
35 -1.384637       0.3187   -0.03705  ...    -0.301543 -0.057227  0.043675   

       Fence  MiscFeature   MiscVal    MoSold    YrSold  SaleType  \
26  0.470579      0.17421 -0.092244 -1.875504  1.713905  0.324245   
27  0.470579      0.17421 -0.092244 -0.405725  1.713905  0.324245   
28  0.470579      0.17421 -0.092244 -0.773170  1.713905 -0.899748   
29  0.470579      0.17421 -0.092244 -0.405725  1.713905  0.324245   
30  0.470579      0.17421 -0.092244 -0.773170  1.713905  0.324245   
31  0.470579      0.17421 -0.092244 -0.773170  1.713905  0.324245   
32 -3.166903      0.17421 -0.092244 -0.773170  1.713905  0.324245   
33  0.470579      0.17421 -0.092244 -0.038281  1.713905  0.324245   
34  0.470579      0.17421 -0.092244 -0.038281  1.713905  0.324245   
35  0.470579      0.17421 -0.092244 -1.140614  1.713905  0.324245   

    SaleCondition  
26       0.201383  
27       0.201383  
28       1.149184  
29       0.201383  
30       0.201383  
31       0.201383  
32      -3.589818  
33       0.201383  
34       0.201383  
35       0.201383  

[10 rows x 80 columns]

In [ ]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   int64  
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   int64  
 6   Alley          1460 non-null   int64  
 7   LotShape       1460 non-null   int64  
 8   LandContour    1460 non-null   int64  
 9   Utilities      1460 non-null   int64  
 10  LotConfig      1460 non-null   int64  
 11  LandSlope      1460 non-null   int64  
 12  Neighborhood   1460 non-null   int64  
 13  Condition1     1460 non-null   int64  
 14  Condition2     1460 non-null   int64  
 15  BldgType       1460 non-null   int64  
 16  HouseStyle     1460 non-null   int64  
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [ ]:
X = scaled_dt.iloc[:, :-1]
y = scaled_dt.iloc[:, -1]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size= 0.33, random_state = 69)

print(f'Training set--> X_train shape= {X_train.shape}, y_train shape= {y_train.shape}')
print(f'Holdout set--> X_test shape= {X_test.shape}, y_test shape= {y_test.shape}')

Training set--> X_train shape= (751, 80), y_train shape= (751,)
Holdout set--> X_test shape= (370, 80), y_test shape= (370,)


##Root Mean Square Error (RMSE) 

In [ ]:
models_scores = [] # To store model scores

def rmse(model):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    
    return mean_squared_error(y_test, y_pred, squared= False) # squared= False > returns Root Mean Square Error                  

#MODEL REGRESSION

##LINEAR REGRESSION

In [ ]:
linear_regression = make_pipeline(LinearRegression())
score = rmse(linear_regression)

models_scores.append(['LinearRegression', score])
print(f'LinearRegression Score= {score}')

LinearRegression Score= 0.4084983403587205


##LASSO REGRESSION

In [ ]:
lasso = make_pipeline(RobustScaler(), Lasso(alpha=0.0005, random_state= 69))

score = rmse(lasso)
models_scores.append(['Lasso', score])
print(f'Lasso Score= {score}')

Lasso Score= 0.40658369354154283


##ELASTICNET REGRESSION

In [ ]:
elastic_net = make_pipeline(RobustScaler(), ElasticNet(alpha=0.0005, l1_ratio= .9, random_state= 69))

score = rmse(elastic_net)
models_scores.append(['ElasticNet', score])
print(f'ElasticNet Score= {score}')

ElasticNet Score= 0.40673613699060196


##KERNEL RIDGE REGRESSION

In [ ]:
kernel_ridge= KernelRidge(alpha=0.6, kernel='polynomial', degree=2, coef0=2.5)
score = rmse(kernel_ridge)
models_scores.append(['KernelRidge', score])
print(f'KernelRidge Score= {score}')

KernelRidge Score= 0.334344647197748


In [ ]:
# Ranking the scores of each model
pd.DataFrame(models_scores).sort_values(by=[1], ascending=True)

0         1
3       KernelRidge  0.334345
1             Lasso  0.406584
2        ElasticNet  0.406736
0  LinearRegression  0.408498

#ENSEMBLE LEARNING

##BAGGING REGRESSOR

In [ ]:
def bagging_predictions(estimator):
    """
    I/P
    estimator: The base estimator from which the ensemble is grown.
    O/P
    br_y_pred: Predictions on test data for the base estimator.
    
    """
    regr = BaggingRegressor(base_estimator=estimator,
                            n_estimators=10,
                            max_samples=1.0,
                            bootstrap=True, # Samples are drawn with replacement
                            n_jobs= n_jobs,
                            random_state=random_state).fit(X_train, y_train)

    br_y_pred = regr.predict(X_test)

    rmse_val = mean_squared_error(y_test, br_y_pred, squared= False) # squared= False > returns Root Mean Square Error   

    print(f'RMSE for base estimator {regr.base_estimator_} = {rmse_val}\n')
    return br_y_pred


predictions = np.column_stack((bagging_predictions(linear_regression),
                              bagging_predictions(lasso),
                              bagging_predictions(elastic_net),
                              bagging_predictions(kernel_ridge)))

print(f"Bagged predictions shape: {predictions.shape}")


RMSE for base estimator Pipeline(steps=[('linearregression', LinearRegression())]) = 0.38501009367911854

RMSE for base estimator Pipeline(steps=[('robustscaler', RobustScaler()),
                ('lasso', Lasso(alpha=0.0005, random_state=69))]) = 0.3841192242234351

RMSE for base estimator Pipeline(steps=[('robustscaler', RobustScaler()),
                ('elasticnet',
                 ElasticNet(alpha=0.0005, l1_ratio=0.9, random_state=69))]) = 0.3841917960817502

RMSE for base estimator KernelRidge(alpha=0.6, coef0=2.5, degree=2, kernel='polynomial') = 0.3012550899563674

Bagged predictions shape: (370, 4)


##FOREST OF RANDOMIZED TREES

###Random Forest Regressor

In [ ]:
RandForestRegressor= RandomForestRegressor(n_estimators=3000,
                                   max_depth=4, max_features='sqrt',
                                   min_samples_leaf=15, min_samples_split=10, 
                                   random_state = random_state)

score = rmse(RandForestRegressor)
models_scores.append(['RandForestRegressor', score])
print(f'RandForestRegressor Score= {score}')

RandForestRegressor Score= 0.4674665197685794


###Extra Trees Regressor

In [ ]:
ExTreesRegressor= ExtraTreesRegressor(n_estimators=3000,
                                   max_depth=4, max_features='sqrt',
                                   min_samples_leaf=15, min_samples_split=10, 
                                   random_state = random_state)

score = rmse(ExTreesRegressor)
models_scores.append(['ExTreesRegressor', score])
print(f'ExTreesRegressor Score= {score}')

ExTreesRegressor Score= 0.5546519797607431


##BOOSTING REGRESSOR

###AddBoost Regressor

In [ ]:
bagging_regressor= BaggingRegressor(n_estimators=3000,
                                   random_state = random_state)

score = rmse(bagging_regressor)
models_scores.append(['BaggingRegressor', score])
print(f'BaggingRegressor Score= {score}')

BaggingRegressor Score= 0.35265827218259316


###Gradient Boosting Regressor

In [ ]:
gradient_boosting_regressor= GradientBoostingRegressor(n_estimators=3000, learning_rate=0.05,
                                   max_depth=4, max_features='sqrt',
                                   min_samples_leaf=15, min_samples_split=10, 
                                   loss='huber', random_state = random_state)

score = rmse(gradient_boosting_regressor)
models_scores.append(['GradientBoostingRegressor', score])
print(f'GradientBoostingRegressor Score= {score}')

GradientBoostingRegressor Score= 0.3018886773329238


###XGB Regressor

In [ ]:
xgb_regressor= xgb.XGBRegressor(colsample_bytree=0.4603, gamma=0.0468, 
                             learning_rate=0.05, max_depth=3, 
                             min_child_weight=1.7817, n_estimators=3000,
                             reg_alpha=0.4640, reg_lambda=0.8571,
                             subsample=0.5213,verbosity=0, nthread = -1, random_state = random_state)
score = rmse(xgb_regressor)
models_scores.append(['XGBRegressor', score])
print(f'XGBRegressor Score= {score}')

XGBRegressor Score= 0.3136351863648073


###ADB Regressor

In [ ]:
adb_regressor= AdaBoostRegressor(n_estimators=3000,learning_rate=0.05,loss='linear',
                                 random_state=None)

score = rmse(adb_regressor)
models_scores.append(['ADBRegressor', score])
print(f'ADBRegressor Score= {score}')

ADBRegressor Score= 0.47401049984610794


##Voting Regressor

In [ ]:
estimators = [ ('elastic_net', elastic_net), ('kernel_ridge', kernel_ridge),('xgb_regressor', xgb_regressor) ]

VoteRegressor = VotingRegressor(estimators=estimators)

score = rmse(VoteRegressor)
models_scores.append(['VoteRegressor', score])
print(f'VoteRegressor Score= {score}')

VoteRegressor Score= 0.31700111439583556


##Stacking Regressor

In [ ]:
estimators = [ ('elastic_net', elastic_net), ('kernel_ridge', kernel_ridge),('xgb_regressor', xgb_regressor) ]

stack = StackingRegressor(estimators=estimators, final_estimator= lasso, cv= 5, n_jobs= n_jobs, passthrough = True)

stack.fit(X_train, y_train)

pred = stack.predict(X_test)

rmse_val = mean_squared_error(y_test, pred, squared= False) # squared= False > returns Root Mean Square Error    
models_scores.append(['Stacking', rmse_val])
print(f'rmse= {rmse_val}')

rmse= 0.45685125197680565


In [ ]:
# Ranking the scores of each model
pd.DataFrame(models_scores).sort_values(by=[1], ascending=True)

0         1
7   GradientBoostingRegressor  0.301889
8                XGBRegressor  0.313635
10              VoteRegressor  0.317001
3                 KernelRidge  0.334345
6            BaggingRegressor  0.352658
1                       Lasso  0.406584
2                  ElasticNet  0.406736
0            LinearRegression  0.408498
11                   Stacking  0.456851
4         RandForestRegressor  0.467467
9                ADBRegressor  0.474010
5            ExTreesRegressor  0.554652

In [ ]:
result = gradient_boosting_regressor.predict(scaled_ds)
print(result)

[ 2.1681417   0.74591317  0.40083533 ... -1.3056809  -0.18431273
  0.64157112]


In [ ]:
result1 = gradient_boosting_regressor.predict(ds)
result1 = result1.reshape(-1,1)
print(result1)

[[2.5979139 ]
 [2.24809633]
 [2.56722649]
 ...
 [2.38028046]
 [2.29640218]
 [2.59014909]]


#Analisis

Hasil predikisi dari masing masing model regresi tidak memiliki perbedaan yang jauh secara significant(similar). linear, Lasso dan Elastic Regression memiliki hasil yang hampir sama yaitu pada angka 0.40 sedangkan untuk Kernel Rridge Regression terlihat berbeda dengan hasil prediksinya yang cukup rendah dibandingkan ketiga model regression tersebut. Hal ini berbeda dengan ensamble learning yang mana hasil prediksinya berbeda satu sama lain. Hasil prediksi yang paling akurat adalah pada model Extra Trees Regressor dengan hasil prediksi mendekati angka 0.50 sedangkan model ensamble learning yang paling tidak akurat yaitu Gradient Boosting Regreesor dengan hasil prediksi 0.30